**Creation Date:**

**April 4, 2020**

**created by Engin Deniz ALPMAN**

In [1]:
import datetime

## 1) Naive datetime

### date Object

This does not have a hour,minute,secons,microseconds

In [5]:
d = datetime.date(2017, 07, 24)

SyntaxError: invalid token (<ipython-input-5-6b237c550810>, line 1)

In [6]:
d = datetime.date(2017, 7, 24)

In [7]:
print(d)

2017-07-24


#### Getting Today date Object

In [9]:
tday = datetime.date.today()

In [10]:
print(tday)

2020-04-03


#### Accessing Basic Attributes of datetime Object

In [11]:
d.day

24

In [12]:
d.year

2017

In [13]:
d.month

7

#### Getting Day of the Week

##### weekday()

Monday: 0 ...... Sunday: 6

In [14]:
d.weekday()

0

##### isoweekday()

Monday:1 ....... Sunday:7

In [15]:
d.isoweekday()

1

### timedelta()

In [16]:
tdelta = datetime.timedelta(days=7)

In [20]:
print(tday)

2020-04-03


In [18]:
print(tday + tdelta)

2020-04-10


In [21]:
print(tday - tdelta)

2020-03-27


**And if we add or subtract day from a date we will get timedelta object**

In [22]:
d - tday

datetime.timedelta(days=-984)

In [23]:
type(d - tday)

datetime.timedelta

**So if we add or subtract timedelta from date, we will get date as a result and if we subtract or add date from a date we will get timedelta**

In [25]:
# Calculating how many days are there until birtday

bday = datetime.date(2020, 6, 1)

In [29]:
count = bday - tday

print(count)

59 days, 0:00:00


In [32]:
# printing just a date
print(count.days)

59


#### Total Seconds

In [33]:
print(count.total_seconds())

5097600.0


### time Object

This does not have year,month,day

Hours,Minutes,Seconds,Microseconds

In [36]:
t = datetime.time(10,40,49,1000)

In [37]:
print(t.hour)

10


In [38]:
print(t.minute)

40


### datetime Object

It has information of both date and time objects

year,month,day,minutes,seconds,microseconds

In [53]:
dt = datetime.datetime(2020, 4, 24, 10, 40, 49, 1000)

In [54]:
print(dt.day)

24


In [55]:
print(dt.year)

2020


In [56]:
print(dt.second)

49


#### Accessing just a date Attribute

In [57]:
print(dt.date())

2020-04-24


####  Accessing just a time Attribute

In [58]:
print(dt.time())

10:40:49.001000


#### Adding timedelta to datetime Object

In [59]:
?datetime.timedelta

In [61]:
tdelta = datetime.timedelta(days=7,seconds=10)

In [62]:
print(dt + tdelta)

2020-05-01 10:40:59.001000


### datetime.today() vs datetime.now() vs datetime.utcnow()

In [63]:
dt_today = datetime.datetime.today()
dt_now = datetime.datetime.now()
dt_utcnow = datetime.datetime.utcnow()

In [65]:
print(dt_today)

2020-04-03 23:02:24.401844


In [66]:
print(dt_now)

2020-04-03 23:02:24.401844


In [67]:
print(dt_utcnow)

2020-04-03 20:02:24.401844


* **datetime.datetime.today() returns a local datetime with a time zone of None**

* **datetime.datetime.now() gives us an option to give time zone**

In [69]:
?datetime.datetime.now

* **datetime.datetime.utcnow() will give us utc time but it is not time zone aware datetime, tz attribute will be empty it will just corresponding utc time**

## 2) Time Zone Aware datetime

we will use pytz package for this

In [71]:
import pytz

In [72]:
dt = datetime.datetime(2020, 4, 24, 10, 40, 49, tzinfo = pytz.UTC)

In [73]:
print(dt)

2020-04-24 10:40:49+00:00


### Creating Current UTC time

#### With datetime.datetime.now()

In [74]:
dt_now = datetime.datetime.now(tz=pytz.UTC)

In [75]:
print(dt_now)

2020-04-03 20:41:43.297589+00:00


#### With datetime.datetime.utcnow()

In [84]:
dt_utcnow = datetime.datetime.utcnow().replace(tzinfo=pytz.UTC)

In [85]:
print(dt_utcnow)

2020-04-03 20:47:08.249508+00:00


### Converting Time Zone

In [86]:
dt_ist = dt_utcnow.astimezone(pytz.timezone("Asia/Istanbul"))

In [87]:
print(dt_ist)

2020-04-03 23:47:08.249508+03:00


2020-04-03 23:47:08.249508+**03:00** that is the hours difference between specified time zone and UTC (in this example specified time zone is 3 hours ahead UTC)

In [89]:
# viewing time zones
pytz.all_timezones

['Africa/Abidjan',
 'Africa/Accra',
 'Africa/Addis_Ababa',
 'Africa/Algiers',
 'Africa/Asmara',
 'Africa/Asmera',
 'Africa/Bamako',
 'Africa/Bangui',
 'Africa/Banjul',
 'Africa/Bissau',
 'Africa/Blantyre',
 'Africa/Brazzaville',
 'Africa/Bujumbura',
 'Africa/Cairo',
 'Africa/Casablanca',
 'Africa/Ceuta',
 'Africa/Conakry',
 'Africa/Dakar',
 'Africa/Dar_es_Salaam',
 'Africa/Djibouti',
 'Africa/Douala',
 'Africa/El_Aaiun',
 'Africa/Freetown',
 'Africa/Gaborone',
 'Africa/Harare',
 'Africa/Johannesburg',
 'Africa/Juba',
 'Africa/Kampala',
 'Africa/Khartoum',
 'Africa/Kigali',
 'Africa/Kinshasa',
 'Africa/Lagos',
 'Africa/Libreville',
 'Africa/Lome',
 'Africa/Luanda',
 'Africa/Lubumbashi',
 'Africa/Lusaka',
 'Africa/Malabo',
 'Africa/Maputo',
 'Africa/Maseru',
 'Africa/Mbabane',
 'Africa/Mogadishu',
 'Africa/Monrovia',
 'Africa/Nairobi',
 'Africa/Ndjamena',
 'Africa/Niamey',
 'Africa/Nouakchott',
 'Africa/Ouagadougou',
 'Africa/Porto-Novo',
 'Africa/Sao_Tome',
 'Africa/Timbuktu',
 'Africa/

### Making Naive datetime Time Zone Aware

#### Fist Way

In [108]:
dt_ist = datetime.datetime.now()

In [109]:
# this is not time zone aware
print(dt_ist)

2020-04-03 23:59:59.289333


In [110]:
dt_ist = dt_ist.astimezone(pytz.timezone("Asia/Istanbul"))

In [112]:
print(dt_ist)

2020-04-03 23:59:59.289333+03:00


#### Second Way

In [116]:
dt_ist = datetime.datetime.now()
print(dt_ist)

2020-04-04 00:06:09.513323


In [117]:
ist_tz = pytz.timezone("Asia/Istanbul")

In [118]:
dt_ist = ist_tz.localize(dt_ist)

In [119]:
print(dt_ist)

2020-04-04 00:06:09.513323+03:00


In [127]:
print(dt_ist.tzinfo)

Asia/Istanbul


### Displaying datetime

In [131]:
dt_ist = dt_utcnow.astimezone(pytz.timezone("Asia/Istanbul"))

In [132]:
print(dt_ist.isoformat())

2020-04-03T23:47:08.249508+03:00


#### Formatting datetime

You don't have to memorize this, you can always check documentation

In [133]:
print(dt_ist.strftime("%B %d, %Y"))

April 03, 2020


### Converting String to datetime

We will use strptime

In [134]:
print(dt_ist.strftime("%B %d, %Y"))

April 03, 2020


In [135]:
dt_str = "April 03, 2020"

In [136]:
dt = datetime.datetime.strptime(dt_str, "%B %d, %Y")

In [137]:
print(dt)

2020-04-03 00:00:00


* **strftime -> Converts Datetime to String**
* **strptime -> Converts String to Datetime**